pkg/daemon/certs/certs.go

```diff
@@ -26,26 +26,63 @@ import (
 	"errors"
 	"fmt"
 	"os"
+	"reflect"
 	"syscall"
 	"unsafe"
 
 	"golang.org/x/sys/windows"
 	"k8s.io/klog/v2"
 )
 
-// Reconcile ensures the certificates in the given trusted CA bundle file are imported as system certificates.
+// previousCABundleFileName tracks all the certs that the operator has imported into the node's local trust store
+const previousCABundleFileName = "imported-certs.pem"
+
+// Reconcile ensures the node's system certificates match the expected state as given by the trusted CA bundle file.
 // Returns a boolean whether any certs needed to be reconciled
-func Reconcile(caBundle string) (bool, error) {
-	if caBundle == "" {
-		// TODO: ensure that all certs that we've added previously are removed https://issues.redhat.com/browse/WINC-688
-		return false, nil
+func Reconcile(caBundlePath string) (bool, error) {
+	var expectedCerts []*x509.Certificate
+	var err error
+	// Read expected certs from CA trust bundle file. Will be empty when a cluster-wide proxy is not in use or removed.
+	if caBundlePath != "" {
+		if expectedCerts, err = readCertsFromFile(caBundlePath); err != nil {
+			return false, fmt.Errorf("Failed to read certificate file %s: %w", caBundlePath, err)
+		}
+		klog.Infof("read %d expected Certs from %s", len(expectedCerts), caBundlePath)
+	}
+
+	// Populate existing certs using the previously imported CA trust bundle.
+	pwd, err := os.Getwd()
+	if err != nil {
+		return false, err
+	}
+	previousCABundlePath := fmt.Sprintf("%s\\%s", pwd, previousCABundleFileName)
+	klog.Infof("previousCABundlePath: %s", previousCABundlePath)
+	exists, err := fileExists(previousCABundlePath)
+	if err != nil {
+		return false, err
+	}
+	// The file will not exist on first reconcile in which case existingCerts will be empty.
+	var existingCerts []*x509.Certificate
+	if exists {
+		if existingCerts, err = readCertsFromFile(previousCABundlePath); err != nil {
+			return false, fmt.Errorf("Failed to read certificate file %s: %w", previousCABundlePath, err)
+		}
 	}
-	// Read expected certs from CA trust bundle file
-	expectedCerts, err := readCertsFromFile(caBundle)
+
+	certChange, err := ensureCertsAreUpToDate(expectedCerts, existingCerts)
 	if err != nil {
-		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+		return false, err
+	}
+
+	// Update previously imported CA trust bundle after all certs have been reconciled
+	if caBundlePath != "" {
+		err = copyFile(caBundlePath, previousCABundlePath)
+		klog.Infof("copied file contents from %s to %s", caBundlePath, previousCABundlePath)
+	} else {
+		// To avoid leaking certs, ensure the imported CA trust bundle file is removed when proxy is removed
+		err = ensureFileIsRemoved(previousCABundlePath)
 	}
-	return ensureCertsAreImported(expectedCerts)
+	return certChange, err
```

Comment:
How do we recover from the errors here?

Reply:
An error here would imply that the `imported-certs.pem` didn't get updated. Since `imported-certs.pem` is one of the sources of truth for determining `certChange`, next reconcile the error would self-correct properly.

Reply:
Please call this out in a comment. It is not immediately obvious.